File: src/core/detector.ts
import { SourceType } from './types';

export function detectSourceType(rawText: string, cleanedText?: string): SourceType {
  const lowerRaw = rawText.toLowerCase();
  const textToAnalyze = cleanedText || rawText;
  const lowerCleaned = textToAnalyze.toLowerCase();
  
  // GitHub PR
  // We check raw text for these because UI menus often contain them, 
  // but they are strong indicators of the page type.
  const isGitHubPR = 
    (lowerRaw.includes('pull request') || lowerRaw.includes('files changed') || lowerRaw.includes('commits')) &&
    (lowerRaw.includes('requested changes') || lowerRaw.includes('approved') || lowerRaw.includes('review') || lowerRaw.includes('merge pull request'));
    
  if (isGitHubPR) {
    return 'github_pr';
  }

  // GitHub Issue
  const isGitHubIssue = 
    lowerRaw.includes('issue') &&
    (lowerRaw.includes('labels') || lowerRaw.includes('assignees') || lowerRaw.includes('milestone') || lowerRaw.includes('open issue'));
    
  if (isGitHubIssue) {
    return 'github_issue';
  }

  // Chat (Slack, Discord, Teams)
  // Heuristic: Many short lines starting with names or timestamps
  // e.g., "John Doe 10:42 AM" or "[10:42] John:"
  const chatMessagePattern = /^\[?\d{1,2}:\d{2}(?:\s*[AP]M)?\]?\s*[\w\s]+:/im;
  const chatNameTimePattern = /^[\w\s]+\s+\d{1,2}:\d{2}\s*[AP]M/im;
  const slackPattern = /new messages?\s*since/im;
  
  const lines = textToAnalyze.split('\n').map(l => l.trim()).filter(l => l.length > 0);
  
  let chatLineCount = 0;
  for (const line of lines) {
    if (chatMessagePattern.test(line) || chatNameTimePattern.test(line)) {
      chatLineCount++;
    }
  }
  if (chatLineCount > 3 || slackPattern.test(lowerRaw)) {
    return 'chat';
  }

  // Docs/Article
  // Heuristic: Check the cleaned text for paragraph density and headings
  // We consider a "long line" as a proxy for a paragraph if lines aren't wrapped
  const longLines = lines.filter(l => l.length > 80).length;
  const headings = (textToAnalyze.match(/^#+\s/gm) || []).length;
  const listItems = (textToAnalyze.match(/^[-*]\s/gm) || []).length;
  
  if (longLines > 3 && headings > 0) {
    // Distinguish between docs and article
    if (listItems > longLines * 2 || lowerCleaned.includes('api') || lowerCleaned.includes('documentation') || lowerCleaned.includes('tutorial')) {
      return 'docs';
    }
    return 'article';
  }
  
  if (longLines > 5) {
    return 'article';
  }

  return 'generic';
}

File: android/app/src/main/java/com/voku/textcleaner/core/Detector.kt
package com.voku.textcleaner.core

/**
 * Heuristic source-type detector.
 * Direct port of `src/core/detector.ts`.
 */
object Detector {

    private const val MIN_CHAT_MATCHES = 3
    private const val MIN_CHAT_TOTAL_LINES = 5

    private val chatMessagePattern =
        Regex("^\\[?\\d{1,2}:\\d{2}(?:\\s*[AP]M)?]?\\s*[\\w\\s]+:", setOf(RegexOption.IGNORE_CASE, RegexOption.MULTILINE))
    private val chatNameTimePattern =
        Regex("^[\\w\\s]+\\s+\\d{1,2}:\\d{2}\\s*[AP]M", setOf(RegexOption.IGNORE_CASE, RegexOption.MULTILINE))
    private val slackPattern =
        Regex("new messages?\\s*since", setOf(RegexOption.IGNORE_CASE, RegexOption.MULTILINE))

    private val headingPattern = Regex("^#+\\s", RegexOption.MULTILINE)
    private val listItemPattern = Regex("^[-*]\\s", RegexOption.MULTILINE)

    private val docsKeywords = listOf("api", "documentation", "tutorial", "guide", "reference", "install", "configuration")
    private val articleKeywords = listOf("article", "analysis", "opinion", "story", "interview", "essay")

    fun detectSourceType(rawText: String, cleanedText: String? = null): SourceType {
        val lowerRaw = rawText.lowercase()
        val textToAnalyze = cleanedText ?: rawText
        val lowerCleaned = textToAnalyze.lowercase()

        // GitHub PR
        val isGitHubPR =
            (lowerRaw.contains("pull request") || lowerRaw.contains("files changed") || lowerRaw.contains("commits")) &&
            (lowerRaw.contains("requested changes") || lowerRaw.contains("approved") || lowerRaw.contains("review") || lowerRaw.contains("merge pull request"))

        if (isGitHubPR) {
            return SourceType.GITHUB_PR
        }

        // GitHub Issue
        val isGitHubIssue =
            lowerRaw.contains("issue") &&
            (lowerRaw.contains("labels") || lowerRaw.contains("assignees") || lowerRaw.contains("milestone") || lowerRaw.contains("open issue"))

        if (isGitHubIssue) {
            return SourceType.GITHUB_ISSUE
        }

        // Chat (Slack, Discord, Teams)
        val lines = textToAnalyze.split("\n").map { it.trim() }.filter { it.isNotEmpty() }

        var chatLineCount = 0
        for (line in lines) {
            if (chatMessagePattern.containsMatchIn(line) || chatNameTimePattern.containsMatchIn(line)) {
                chatLineCount++
            }
        }
        if ((chatLineCount >= MIN_CHAT_MATCHES && lines.size >= MIN_CHAT_TOTAL_LINES) || slackPattern.containsMatchIn(lowerRaw)) {
            return SourceType.CHAT
        }

        // Docs/Article heuristics
        val longLines = lines.count { it.length > 80 }
        val paragraphLines = lines.count { it.length > 40 }
        val headings = headingPattern.findAll(textToAnalyze).count()
        val listItems = listItemPattern.findAll(textToAnalyze).count()
        val docsKeywordMatches = docsKeywords.count { lowerCleaned.contains(it) }
        val articleKeywordMatches = articleKeywords.count { lowerCleaned.contains(it) }

        if ((headings > 0 && listItems >= 2) || docsKeywordMatches >= 2) {
            return SourceType.DOCS
        }

        if ((headings > 0 && paragraphLines >= 2) || longLines > 3) {
            // Distinguish between docs and article
            if (listItems > 2 || docsKeywordMatches >= articleKeywordMatches + 1) {
                return SourceType.DOCS
            }
            return SourceType.ARTICLE
        }

        if (paragraphLines >= 3 || articleKeywordMatches > 0) {
            return SourceType.ARTICLE
        }

        return SourceType.GENERIC
    }
}
